### PR TITLE
Make fast exit algorithm optional for AMD

### PR DESCRIPTION
**Pull request by jean-m-cyr**
- Add --cl-noexit parameter to disable fast exit
- Modify opencl ethash to support no exit
- Create noexit versions of binary kernels

### CODE REVIEW

File: ethminer/main.cpp [hunk 1]
```diff
@@ -320,6 +320,8 @@ class MinerCLI
 
         app.add_flag("--cl-nobin", m_CLSettings.noBinary, "");
 
+        app.add_flag("--cl-noexit", m_CLSettings.noExit, "");
```
Comment (chfast) [inline ethminer/main.cpp]: I know there is `--cl-nobin` but can we remove the negation from the option? Make it `--cl-fast-exit`.

File: libethash-cl/kernels/cl/ethash.cl [hunk 2]
```diff
@@ -259,8 +259,10 @@ __kernel void search(
     uint isolate
 )
 {
+#ifndef NO_FAST_EXIT
```
Comment (chfast) [inline libethash-cl/kernels/cl/ethash.cl]: It should be `#ifdef FAST_EXIT` or better `#if FAST_EXIT`.

**Comment by jean-m-cyr** [comment]
Can't really use --cl-fastexit if I want fast exit to be the default. The parser allows flags to be turned on, but not turned off!

File: libethash-cl/kernels/cl/ethash.cl (hunk 2 above)
Comment (jean-m-cyr) [reply libethash-cl/kernels/cl/ethash.cl]: Done

File: libethash-cl/CLMiner.cpp [hunk 3]
```diff
@@ -412,10 +417,12 @@ void CLMiner::workLoop()
             current = w;  // kernel now processing newest work
             current.startNonce = startNonce;
             // Increase start nonce for following kernel execution.
-            startNonce += results.hashCount * m_settings.localWorkSize;
-
+            startNonce += m_settings.globalWorkSize;
```
Comment (chfast) [inline libethash-cl/CLMiner.cpp]: Why this change?

File: libethash-cl/CLMiner.cpp [hunk 4]
```diff
@@ -432,7 +439,7 @@ void CLMiner::kick_miner()
 {
     // Memory for abort Cannot be static because crashes on macOS.
     const uint32_t one = 1;
-    if (m_abortqueue.size())
+    if (m_abortqueue.size() && !m_settings.noExit)
```
Comment (chfast) [inline libethash-cl/CLMiner.cpp]: The flag should probably be checked first. And also maybe `!m_abortqueue.empty()`?

File: libethash-cl/CLMiner.cpp (hunk 3 above)
Comment (jean-m-cyr) [reply libethash-cl/CLMiner.cpp]: Because it is more efficient. Consider the 3 scenarios where we get here:
1 - All the work groups have terminated without finding a solution.
     In this case the number of hashes calculated will always be equal to the global work size.
2 - Some work groups have exited early because a new job has arrived.
     A new job has arrived. It doesn't matter what we increment to, the start hash will be reset
     for the new job anyway.
3 - Some work groups have exited early because a work group has found a solution.
     In this case we may increment past some nonces that were not calculated by the work
     groups that terminated early. This is no consequence since our chances of finding a solution
     depend on the # of nonces checked, that does not depend on the checked nonces being
     consecutive.

File: ethminer/main.cpp (hunk 1 above)
Comment (chfast) [reply ethminer/main.cpp]: > Can't really use --cl-fastexit if I want fast exit to be the default. The cmd. line parser allows flags to be turned on, but not turned off!

File: libethash-cl/kernels/isa/GCN5_ethash_search.isa [hunk 5]
```diff
@@ -850,7 +854,7 @@
         s_waitcnt       vmcnt(0) & lgkmcnt(0)
 
         /* store the solution */
-        v_min_u32       mix_double_temp0[0], 3, temp0
+        v_min_u32       mix_double_temp0[0], MAX_SEARCH_RESULTS, temp0
```
[the remark this answers is not in the document]
Comment (jean-m-cyr) [reply libethash-cl/kernels/isa/GCN5_ethash_search.isa]: The max results for CUDA is 4, and for OpenCL is 15. This should actually read MAX_SEARCH_RESULTS - 1.

**Comment by jean-m-cyr** [comment]
@chfast I'm puzzled! How does one respond to requested changes without committing new changes?

**Comment by jean-m-cyr** [comment]
@chfast I measure no difference with or without isolate. I believe it may have been relevant with OpenCL 1.1 (which I'm not sure we even suport anymore?).

**Comment by chfast** [comment]
> How does one respond to requested changes without committing new changes?

Commit them to another branch if that's the question.
